File: LeetCode/src/com/ejolie/challenge/_2021/february/Day2.java
package com.ejolie.challenge._2021.february;

import com.ejolie.problems.structure.TreeNode;

/**
 * Day2. Trim a Binary Search Tree
 */
public class Day2 {
    public TreeNode trimBST(TreeNode root, int low, int high) {
        if (root == null) {
            return null;
        }

        if (root.val >= low && root.val <= high) {
            root.left = trimBST(root.left, low, high);
            root.right = trimBST(root.right, low, high);
        } else if (root.val < low) {
            return trimBST(root.right, low, high);
        } else { // root.val > high
            return trimBST(root.left, low, high);
        }

        return root;
    }
}

File: LeetCode/src/com/ejolie/challenge/_2021/september/Day2.py
from typing import List, Optional

"""
Day2. Unique Binary Search Trees II
"""
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        return self.makeBST(start=1, end=n)
    
    def makeBST(self, start: int, end: int) -> List[TreeNode]:
        if start > end:
            return [None]
        
        allTrees = []
        for val in range(start, end+1):
            allLeftTrees = self.makeBST(start=start, end=val-1)
            allRightTrees = self.makeBST(start=val+1, end=end)
            
            for leftTree in allLeftTrees:
                for rightTree in allRightTrees:
                    root = TreeNode(val)
                    root.left = leftTree
                    root.right = rightTree
                    
                    allTrees.append(root)
                
        return allTrees
